### DISCUSSION TITLE
What config or plugins are required for neopixel support?

### DISCUSSION THREAD

**Discussion by dresco** [grblHAL/STM32F7xx#19] [General]
I've pulled the neopixel_spi support into the H7 driver with the latest updates. 
As it's using DMA, will need a couple of local changes to cater for cache maintenance & alignment..

Am wondering what configuration I need to actually test it, other than defining the SPI port?
Does it replace the old 3rd party RGB plugin?

Thanks..

**Comment by terjeio** [comment]
> Am wondering what configuration I need to actually test it, other than defining the SPI port?

There are two basic plugins, one that adds configuration of strip lengths and one that implements a Marlin style M150 command that can be used for testing. These are found in the new [assorted plugins repo](https://github.com/grblHAL/Plugins_misc).

> Does it replace the old 3rd party RGB plugin?

No, it only separates out the lower level APIs. The RGB (or status LED plugin) relies on the configuration plugin and can live alongside the M150 plugin - after it has been updated for the recent user M-code core change.

**Reply by dresco** [reply]
Great, thanks. I've got a small WS2812 strip in the post, will do some testing when it arrives...

**Comment by dresco** [comment]
Have bumped into an interesting 'feature' of the H7 chip while looking at this (from searching online, seems it is the same on the G0 parts).

At the start of a SPI transfer, if the first bit is 1, MOSI is driven high before the clock is enabled. (Have also seen MOSI delayed high at the end of  a transfers in TI mode). See attached image, transferring binary `10100101` in Motorola mode;

![image](https://github.com/user-attachments/assets/68cde231-c345-4a20-b31d-b6ba9c2c8da4)

This isn't an issue for real SPI data, as should only be sampled on the clock edge, but for generating neopixel timing it means the first bit of data is always interpreted as a longer 'high' level.

I need to do a bit more testing, but I think the minimum workaround would be to shift the 3 bit pattern once to the right, from 110/100 to 011/010. By my reckoning, that should still generate the same data stream timing, but always starting with a 0 bit?

The bit stream always ends with 0 anyway, due to the 24 bytes of padding added to `neopixel.num_bytes`. Btw, why is that there, to ensure a valid reset timing between writes?

As an aside, I've seen that some implementations (Zephyr for example), use 8 SPI bits per pulse, allowing the code to specify the high/low bit patterns. This does use more memory, but has the advantage that the bit pattern can be tweaked to better suit available SPI rates. Might be important on STM32 at least where you can only half/double the prescaler rates? Just something to consider..

**Reply by terjeio** [reply]
Bit shifting should work and yes the padding is to ensure the reset timing requirement is met.
Using a 8 bit bit pattern is surely an option, the higher level code does not care about how the data is transferred. If you implement it then consider adding the pack/unpack code to [grbl/rgb.h](https://github.com/grblHAL/core/blob/master/rgb.h).

FYI I have not put too much time into how the data is transferred since, as I understand it, the LEDs are quite forgiving about the bit timing. And LED strip support not a critical part of grblHAL anyway.

**Reply by dresco** [reply]
Cool thanks, I'll submit a PR for the 3-bit packing change.

I'll look into adding an 8-bit option if it becomes necessary. (SPI is currently clocked at 48MHz for all the H7 boards, which works fine for the 3MHz neopixel output).

**Comment by dresco** [comment]
I've [updated](https://github.com/dresco/STM32H7xx/commit/2114999dfd7deffd8995f9478bef9b63825c4a5a) the neopixel SPI code for the H7 driver, all seems to be working as expected now. FYI - there were a few notable differences from the F7 driver.

- Added some code to address alignment & dcache maintenance (expected).
- Had to add the SCK GPIO definitions, else the EOT SPI interrupt would never fire. (Interrupt needed to signal the end of the SPI transaction when using DMA, else would be stuck in busy state after first invocation).
- Added additional macros to simplify the DMA setup. The H7 has a DMA MUX, so can use the same DMA stream, regardless of SPI port.

Have definitely learnt some things along the way, so I will have another look at the WizNet SPI support..
